### PR TITLE
feat(streams): Use Cursor

### PR DESCRIPTION
**Pull request by cynicaljoy**
### Description

Provide support for the new per-event stream cursor.

### Motivation and context

There's an edge case where resuming a stream from a txn time with lots of documents written could result in event loss. This approach fixes that edge case.

### How was the change tested?

Updated integration test

### Change types

* - [x] Bug fix (non-breaking change that fixes an issue)
* - [x] New feature (non-breaking change that adds functionality)
* - [ ] Breaking change (backwards-incompatible fix or feature)

### Checklist:

* - [x] My code follows the code style of this project.
* - [ ] My change requires a change to Fauna documentation.
* - [ ] My change requires a change to the README, and I have updated it accordingly.

----
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Fauna/Types/Stream.cs [hunk 1]
```diff
@@ -28,6 +30,10 @@ public void Serialize(System.IO.Stream stream)
         {
             writer.WriteNumber("start_ts", StartTs.Value);
         }
+        if (LastCursor != null)
```
Comment (pnwpedro) [inline Fauna/Types/Stream.cs:33]: I think we should only be writing either the `start_ts` or the `cursor`, but not both. The `cursor` probably takes priority.

**Review by pnwpedro — CHANGES_REQUESTED** [review]

File: Fauna/Types/Event.cs [hunk 2]
```diff
@@ -20,6 +20,7 @@ public class Event<T> where T : notnull
 {
     public EventType Type { get; private init; }
     public long TxnTime { get; private init; }
+    public string Cursor { get; private init; } = null!;
```
Comment (adambollen) [inline Fauna/Types/Event.cs:23]: I'm not familiar with the `= null!` syntax - what does this do?

Comment (cynicaljoy) [reply Fauna/Types/Event.cs:23]: Rider suggestion - 

image

File: Fauna/Core/StreamOptions.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
+namespace Fauna;
+
+/// <summary>
+/// Represents the options when subscribing to Fauna Streams.
+/// </summary>
+public class StreamOptions
```
Comment (pnwpedro) [inline Fauna/Core/StreamOptions.cs:6]: Let's add more to this class to control for what's allowed instead of the validation in the Client.

1. You can remove { init; }
2. Add constructors that take only the combos allowed, e.g. StreamOptions(token), StreamOptions(token, cursor), StreamOptions(token, startTs)
3. Make the properties readonly

File: Fauna/IClient.cs [hunk 4]
```diff
@@ -520,21 +521,42 @@ internal abstract IAsyncEnumerator<Event<T>> SubscribeStreamInternal<T>(
...
+
+            stream = response.Data;
+        }
+        else
+        {
+            stream = new Stream(streamOptions.Token!);
+        }
+
+        stream.LastCursor = streamOptions?.Cursor;
+        stream.StartTs = streamOptions?.StartTs;
```
Comment (pnwpedro) [inline Fauna/IClient.cs:555]: this doesn't seem right. i'd be surprised if we should be setting these in the case where we're issuing a new query.

File: Fauna/IClient.cs [hunk 5]
```diff
@@ -520,21 +521,42 @@ internal abstract IAsyncEnumerator<Event<T>> SubscribeStreamInternal<T>(
...
+            var response = await QueryAsync<Stream>(
+                query,
+                queryOptions,
+                cancellationToken);
+
+            stream = response.Data;
+        }
+        else
+        {
+            stream = new Stream(streamOptions.Token!);
```
Comment (pnwpedro) [inline Fauna/IClient.cs:551]: there should probably be a constructor that takes token, cursor and start ts in some combination.

also do you need to assert that Token isn't null here? i thought your previous if branch is checking that

File: Fauna/Types/Event.cs [hunk 6]
```diff
@@ -54,6 +56,17 @@ private static long GetTxnTime(JsonElement json)
...
     }
 
+    private static string GetCursor(JsonElement json)
+    {
+        if (!json.TryGetProperty(CursorFieldName, out var elem))
+        {
+            throw new Exception($"Missing required field: cursor - {json.ToString()}");
+        }
+
+        return elem.Deserialize<string>() ?? throw new InvalidOperationException();
```
Comment (pnwpedro) [inline Fauna/Types/Event.cs:66]: why are we throwing InvalidOperationException?

File: Fauna/Types/Event.cs [hunk 7]
```diff
@@ -54,6 +56,17 @@ private static long GetTxnTime(JsonElement json)
         return elem.TryGetInt64(out long i) ? i : default;
     }
 
+    private static string GetCursor(JsonElement json)
+    {
+        if (!json.TryGetProperty(CursorFieldName, out var elem))
+        {
+            throw new Exception($"Missing required field: cursor - {json.ToString()}");
```
Comment (pnwpedro) [inline Fauna/Types/Event.cs:63]: I see we throw Exception elsewhere in this class, but it's too generic.

File: Fauna/Core/Connection.cs [hunk 8]
```diff
@@ -57,44 +60,63 @@ public async IAsyncEnumerable<Event<T>> OpenStream<T>(
...
 
-                while (!streamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
-                {
-                    string? line = await streamReader.ReadLineAsync().WaitAsync(cancellationToken);
-                    if (string.IsNullOrWhiteSpace(line))
+                    if (!response.IsSuccessStatusCode)
                     {
-                        continue;
+                        // TRICKY: we need to break the events listener loop
+                        listener.Dispatch(null);
```
Comment (pnwpedro) [inline Fauna/Core/Connection.cs:79]: afaict, this is essentially calling `listener.Close()`, unless you _must_ enqueue a null event?

**Review by pnwpedro — CHANGES_REQUESTED** [review]
Progress! Let me know if you want to chat about any of these.

File: Fauna/Core/Connection.cs (hunk 8 above)
Comment (cynicaljoy) [reply Fauna/Core/Connection.cs:79]: calling Close wasn't breaking the loop, I had to enqueue something to break on

File: Fauna/Types/Event.cs (hunk 6 above)
Comment (cynicaljoy) [reply Fauna/Types/Event.cs:66]: > The exception that is thrown when a method call is invalid for the object's current state.

It was an exception type that seemed to make sense to me, open to suggestions.

File: Fauna/Types/Event.cs (hunk 7 above)
Comment (cynicaljoy) [reply Fauna/Types/Event.cs:63]: how about `InvalidDataException`?

File: Fauna/IClient.cs (hunk 4 above)
Comment (cynicaljoy) [reply Fauna/IClient.cs:555]: this is to account for the retry, we already control the conditional use of `Cursor ||  StartTs` in the StreamOptions constructor so we don't need to if-else it here.

Comment (cynicaljoy) [reply Fauna/IClient.cs:555]: oh I see what you mean now

File: Fauna/Types/Event.cs (hunk 6 above)
Comment (pnwpedro) [reply Fauna/Types/Event.cs:66]: Is it conceptually the same as "Missing required field etc etc"?

File: Fauna/Types/Event.cs (hunk 7 above)
Comment (pnwpedro) [reply Fauna/Types/Event.cs]: that seems reasonable

File: Fauna/Core/Connection.cs (hunk 8 above)
Comment (pnwpedro) [reply Fauna/Core/Connection.cs:79]: what about writing a method called BreakAndClose() that does that instead of hiding something inside Dispatch?

File: Fauna/Types/Event.cs (hunk 6 above)
Comment (cynicaljoy) [reply Fauna/Types/Event.cs:66]: it's more like "Unexpected value"

Comment (pnwpedro) [reply Fauna/Types/Event.cs:66]: what's the unexpected value when elem.Deserialize<string>() coalesces into the throw? like what's actually happening here?

Comment (cynicaljoy) [reply Fauna/Types/Event.cs]: sycn'd with Adam on this -- we added a message to the exception to clear things up. Using a `throw` here makes the most sense because Deserialize can return null and we need to handle that in rare case that Fauna started returning a `cursor` value that couldn't be deserialized to a string.

Comment (pnwpedro) [reply Fauna/Types/Event.cs]: Okay, I think we're getting closer to the why now, which was the original question. It sounds to me like we should just always have a cursor, is that right? So just deserialize and assert it's not null. Their stack trace will point to this line anyway, it's not actionable by the end user. We could throw a protocol exception, but same idea–it's not actionable.

The above assumes when you say "rare case" it's not a rare legitimate case, but something that Fauna should not do.

**Review by pnwpedro — APPROVED** [review]

**Review by adambollen — APPROVED** [review]
\
